File: cpp/Ch 7.OOD/1.Deck Of Cards/card.cpp
#include <string>
#include <vector>
#include <iostream>

class Card
{
private:
	bool available = true;

protected:
	int faceValue = 0;

	Suit *suit_Renamed;

public:
	Card(int c, Suit *s)
	{
		faceValue = c;
		suit_Renamed = s;
	}

	virtual int value() = 0;

	virtual Suit *suit()
	{
		return suit_Renamed;
	}

	
	virtual bool isAvailable()
	{
		return available;
	}

	virtual void markUnavailable()
	{
		available = false;
	}

	virtual void markAvailable()
	{
		available = true;
	}

	virtual void print()
	{
		std::vector<std::wstring> faceValues = {L"A", L"2", L"3", L"4", L"5", L"6", L"7", L"8", L"9", L"10", L"J", L"Q", L"K"};
		std::wcout << faceValues[faceValue - 1];
		switch (suit_Renamed)
		{
		case Club:
			std::wcout << std::wstring(L"c");
			break;
		case Heart:
			std::wcout << std::wstring(L"h");
			break;
		case Diamond:
			std::wcout << std::wstring(L"d");
			break;
		case Spade:
			std::wcout << std::wstring(L"s");
			break;
		}
		std::wcout << std::wstring(L" ");
	}
};

File: Java/Ch 07. Object-Oriented Design/Q7_01_Deck_of_Cards/Card.java
package Q7_01_Deck_of_Cards;

public abstract class Card {
	private boolean available = true;
	
	/* number or face that's on card - a number 2 through 10, 
	 * or 11 for Jack, 12 for Queen, 13 for King, or 1 for Ace 
	 */
	protected int faceValue;
	protected Suit suit;

	public Card(int c, Suit s) {
		faceValue = c;
		suit = s;
	}
	
	public abstract int value();
	
	public Suit suit() { 
		return suit; 
	}
	
	/* returns whether or not the card is available to be given out to someone */
	public boolean isAvailable() {
		return available;
	}
	
	public void markUnavailable() {
		available = false;
	}
	
	public void markAvailable() {
		available = true;
	}
	
	public void print() {
		String[] faceValues = {"A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"};
		System.out.print(faceValues[faceValue - 1]);
		switch (suit) {
		case Club:
			System.out.print("c");
			break;
		case Heart:
			System.out.print("h");
			break;
		case Diamond:
			System.out.print("d");
			break;
		case Spade:
			System.out.print("s");
			break;			
		}
		System.out.print(" ");
	}
}
